ext/odbc/php_odbc.c

```diff
@@ -427,6 +510,9 @@ PHP_RSHUTDOWN_FUNCTION(odbc)
 /* {{{ PHP_MSHUTDOWN_FUNCTION */
 PHP_MSHUTDOWN_FUNCTION(odbc)
 {
+	zend_hash_destroy(&ODBCG(results));
+	zend_hash_destroy(&ODBCG(connections));
```

Comment:
Shouldn't this happen in GSHUTDOWN? Considering they are init in GINIT.

Also if we don't destroy them per request, wouldn't this allow us to also remove the persistent resource type?

Maybe one "generic" way of doing this to check on the persistent flag via a hash_apply function (or whatever the name of those are)

Reply:
> Shouldn't this happen in GSHUTDOWN? Considering they are init in GINIT.

Hmm, great question... I used ext/pgsql as a reference where the same happens.. According to what I could distill from the externals book, the main difference between MINIT and GINIT is this:

> GINIT() is launched before MINIT() for the current process. In case of NTS, that’s all. In case of ZTS, GINIT() will be called additionally for every new thread spawned by the thread library.

So I guess I should rather destroy the hashes in `GSHUTDOWN` if I'm not mistaken...

> Also if we don't destroy them per request, wouldn't this allow us to also remove the persistent resource type?

Yeah, this crossed my mind as well, but I'm not sure about it, as the same was done in case of mysqli and pgsql as well: https://github.com/php/php-src/pull/6791#discussion_r598510723 But we should definitely discuss the feasibility of your idea with the PHPF. :)